onnxruntime/core/framework/session_state.cc

```diff
@@ -228,4 +228,76 @@ const NodeIndexInfo& SessionState::GetNodeIndexInfo() const {
   return *node_index_info_;
 }
 
+// use a cheap way of matching first. if we have multiple entries with this key, we will do the more expensive
+// check of the individual feed/output names
+static int MakeFeedsFetchesManagerCacheKey(const std::vector<std::string>& feed_names,
+                                           const std::vector<std::string>& output_names) {
+  return static_cast<int>(feed_names.size()) << 16 | static_cast<int>(output_names.size());
+};
+
+const FeedsFetchesManager* SessionState::GetFeedsFetchesManager(const std::vector<std::string>& feed_names,
+                                                                const std::vector<std::string>& output_names) const {
+  int key = MakeFeedsFetchesManagerCacheKey(feed_names, output_names);
+  auto num_matches = cached_feeds_fetches_managers_.count(key);
+
+  const FeedsFetchesManager* manager = nullptr;
+
+  switch (num_matches) {
+    case 0: {
+      break;
+    }
+    case 1: {
+      manager = cached_feeds_fetches_managers_.find(key)->second.get();
```

Comment:
Does this not need an equality check (as in the default case below)? Does the context guarantee that if there is one key-match it must be the right one?

Reply:
You're right. I changed it from doing get+create in one method to separating the two to improve const correctness and overlooked this. Is it necessary to support multiple entries with the same number of inputs and outputs? I can't think of a use case where that limitation couldn't be worked around, as you'd have to be swapping optional inputs/outputs between calls to Run to get a mismatch.  

If not, I could just use std::map, and in the cache method validate there's no clash with an existing entry. That relies on the cache call coming before a newly created instance is used, which is what happens in InferenceSession currently. Not perfect, but makes the getter cheap.

Reply:
Re. your side question: not sure if I understood it correctly. But we need to guarantee correctness, regardless of whether it is a common use case or not, right? If we have multiple optional inputs, one call could pass optional input X, and another call could pass optional input Y, and the caching should ensure results are correct in both cases.